src/viam/sdk/services/mlmodel/server.cpp

```diff
@@ -0,0 +1,191 @@
+// limitations under the License.
+
+#include <viam/sdk/services/mlmodel/server.hpp>
+
+#include <viam/sdk/rpc/server.hpp>
+#include <viam/sdk/services/mlmodel/mlmodel.hpp>
+#include <viam/sdk/services/mlmodel/private/proto.hpp>
+
+namespace viam {
+namespace sdk {
+
+MLModelServiceServer::MLModelServiceServer()
+    : MLModelServiceServer(std::make_shared<ResourceManager>()) {}
+
+MLModelServiceServer::MLModelServiceServer(std::shared_ptr<ResourceManager> manager)
+    : ResourceServer(std::move(manager)) {}
+
+void MLModelServiceServer::register_server(std::shared_ptr<Server> server) {
+    server->register_service(this);
+}
+
+::grpc::Status MLModelServiceServer::Infer(
+    ::grpc::ServerContext* context,
+    const ::viam::service::mlmodel::v1::InferRequest* request,
+    ::viam::service::mlmodel::v1::InferResponse* response) {
+    if (!request) {
+        return {::grpc::StatusCode::INVALID_ARGUMENT, "Called [Infer] without a request"};
+    };
+
+    std::shared_ptr<Resource> rb = resource_manager()->resource(request->name());
+
+    if (!rb) {
+        return {::grpc::UNKNOWN, "resource not found: " + request->name()};
+    }
+
+    std::shared_ptr<MLModelService> mlms = std::dynamic_pointer_cast<MLModelService>(rb);
+
+    if (!request->has_input_data()) {
+        return {::grpc::StatusCode::INVALID_ARGUMENT, "Called [Infer] with no input data"};
+    }
+
+    // TODO: Metadata should be cached client side so we don't need to
+    // do two round trips for every inference.
+    const auto md = mlms->metadata();
+
+    mlmodel_details::tensor_storage input_storage;
+    MLModelService::named_tensor_views inputs;
+
+    const auto& input_fields = request->input_data().fields();
+    for (const auto& input : md.inputs) {
+        const auto where = input_fields.find(input.name);
+
+        // Ignore any inputs for which we don't have metadata, since
+        // we can't know what type they should decode to.
+        //
+        // TODO: Should this be an error? For now we just don't decode
+        // those tensors.
+        if (where == input_fields.end()) {
+            continue;
+        }
+
+        auto status =
+            mlmodel_details::pb_value_to_tensor(input, where->second, &input_storage, &inputs);
+        if (!status.ok()) {
+            return status;
+        }
+    }
+
+    // TODO: Should we handle exceptions here? Or is it ok to let them
+    // bubble up and have a higher layer deal with it? Perhaps all of
+    // our server side overrides should be `noexcept`.
+    //
+    // NOTE: Exceptions from here will result in an empty reply. We do
+    // need to handle them and convert into a failing reply status.
+    const auto outputs = mlms->infer(inputs);
+    auto& pb_output_data_fields = *(response->mutable_output_data()->mutable_fields());
+    for (const auto& kv : *outputs) {
+        auto emplace_result = pb_output_data_fields.try_emplace(kv.first);
```

Comment:
I'm going to TODO this for now, since I will need to come back through and validate this entire change against bullseye / focal which will catch this, but I can't do that until I've made this work against vendored xtensor. I'd rather fix up any older proto version incompatibilities in one go.